File: PERL/Cap08/c08ex08.pl
# c08ex08.pl

#!/usr/bin/perl

use strict;
use warnings;

my $FIM = 10;
my @MATRIZ;
my $INICIO;

sub vazia {
  if ($INICIO == 0) {
    return 1;
  } else {
    return 0;
  }
}

sub cheia {
  if ($INICIO == $FIM) {
    return 1;
  } else {
    return 0;
  }
}

sub adicionar {
  my ($ELEMENTO) = @_;
  if (cheia()) {
    return 0;
  } else {
    $INICIO = $INICIO + 1;
    $MATRIZ[$INICIO] = $ELEMENTO;
    return 1;
  }
}

sub retirar {
  my $ELEMENTO = shift;
  if (vazia()) {
    return 0;
  } else {
    $$ELEMENTO = $MATRIZ[1];
    for (my $I = 1; $I < $INICIO; $I++) {
      $MATRIZ[$I] = $MATRIZ[$I + 1];
    }
    $INICIO = $INICIO - 1;
    return 1;
  }
}

sub entrada {
  print "\n";
  print "Entre um elemento numerico: ";
  chomp(my $X = <STDIN>);
  if (adicionar($X)) {
    print "\n";
    print "Elemento $X inserido na posicao $INICIO.\n";
    print "\n";
  } else {
    print "\n";
    print "Impossivel adicionar $X - fila lotada.\n";
    print "\n";
  }
}

sub saida {
  print "\n";
  my $X;
  if (retirar(\$X)) {
    print "Elemento $X retirado do inicio da fila.\n";
    print "\n";
  } else {
    print "Impossivel retirar elemento - fila vazia.\n";
    print "\n";
  }
}

sub atual {
  print "\n";
  if (!vazia()) {
    print $MATRIZ[1] . " - primeiro elemento da fila.\n";
    print "\n";
  } else {
    print "Impossivel apresentar - fila vazia.\n";
    print "\n";
  }
}

sub exibicao {
  print "\n";
  if (!vazia()) {
    for (my $I = 1; $I <= $INICIO; $I++) {
      print "Posicao: $I possui o elemento $MATRIZ[$I].\n";
    }
  } else {
    print "Impossivel apresentar - fila vazia.\n";
  }
  print "\n";
}

sub criar {
  $INICIO = 0;
  for (my $I = 1; $I <= $FIM; $I++) {
    $MATRIZ[$I] = 0;
  }
}

criar();
my $OPCAO = 0;
while ($OPCAO != 6) {
  print "PROGRAMA: FILA\n\n";
  print "[1] - Entrada\n";
  print "[2] - Saida\n";
  print "[3] - Apresentar 1o. da fila\n";
  print "[4] - Apresentar a fila\n";
  print "[5] - Nova fila\n";
  print "[6] - Sair\n\n";

  print "Escolha uma opcao: ";
  chomp($OPCAO = <STDIN>);
  print "\n";

  if ($OPCAO != 6) {
    if ($OPCAO == 1) {
      entrada();
    } elsif ($OPCAO == 2) {
      saida();
    } elsif ($OPCAO == 3) {
      atual();
    } elsif ($OPCAO == 4) {
      exibicao();
    } elsif ($OPCAO == 5) {
      criar();
      print "\n";
    }
  }
}

File: RUBY/Cap08/c08ex08.rb
# c08ex08.rb

#!/usr/bin/env ruby

LIMITE = 10
$matriz = Array.new(LIMITE)

def vazia
  if ($inicio == -1)
    true
  else
    false
  end
end

def cheia
  if ($inicio == LIMITE - 1)
    true
  else
    false
  end
end

def adicionar(elemento)
  if (cheia)
    false
  else
    $inicio += 1
    $matriz[$inicio] = elemento
    true
  end
end

def retirar(elemento)
  if (vazia)
    false
  else
    elemento = $matriz[0]
    for i in 1..$inicio - 1
      $matriz[i] = $matriz[i + 1]
    end
    $inicio -= 1
    true
  end
end

def entrada
  print "Entre um elemento numerico: "
  x = gets.chomp.to_i
  puts "\n"
  if (adicionar(x))
    puts "Elemento #{x} inserido na posicao #{$inicio}."
  else
    puts "Impossivel adicionar #{x} - fila lotada."
  end
  puts
end

def saida
  if (vazia)
    puts "Impossivel retirar elemento - fila vazia."
  else
    elemento = $matriz[1]
    for i in 0..$inicio - 1
      $matriz[i] = $matriz[i + 1]
    end
    $inicio -= 1
    puts "Elemento #{elemento} retirado do inicio da fila."
  end
  puts
end

def atual
  if (!vazia)
    puts "#{$matriz[0]} - primeiro elemento da fila."
  else
    puts "Impossivel apresentar - fila vazia."
  end
  puts
end

def exibicao
  if (!vazia)
    for i in 0..$inicio
      printf "Posicao: %2d: possui o elemento %d.\n", i + 1, $matriz[i]
    end
  else
    puts "Impossivel apresentar - fila vazia."
  end
  puts
end

def criar
  $inicio = 0
  for i in 0..(LIMITE - 1)
    $matriz[i] = 0
  end
end

criar
opcao = 0
while (opcao != 6)
  puts "PROGRAMA: FILA\n\n"
  puts "[1] - Entrada\n"
  puts "[2] - Saida\n"
  puts "[3] - Apresentar 1o. da fila\n"
  puts "[4] - Apresentar a fila\n"
  puts "[5] - Nova fila\n"
  puts "[6] - Sair\n\n"

  print "Escolha uma opcao: "
  opcao = gets.chomp.to_i
  puts "\n"

  if (opcao != 6)
    case (opcao)
    when 1
      entrada
    when 2
      saida
    when 3
      atual
    when 4
      exibicao
    when 5
      criar
      puts
    end
  end
end
